crates/dojo-lang/src/contract.rs

```diff
@@ -197,8 +200,232 @@ impl DojoContract {
             .to_string(),
         )]
     }
-}
 
-// fn is_context(db: &dyn SyntaxGroup, param: &Param) -> bool {
-//     param.type_clause(db).ty(db).as_syntax_node().get_text(db) == "Context"
-// }
+    /// Gets name, modifiers and type from a function parameter.
+    pub fn get_parameter_info(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param: ast::Param,
+    ) -> (String, String, String) {
+        let name = param.name(db).text(db).trim().to_string();
+        let modifiers = param.modifiers(db).as_syntax_node().get_text(db).trim().to_string();
+        let param_type =
+            param.type_clause(db).ty(db).as_syntax_node().get_text(db).trim().to_string();
+
+        (name, modifiers, param_type)
+    }
+
+    /// Check if the function has a self parameter.
+    ///
+    /// Returns
+    ///  * a boolean indicating if `self` has to be added,
+    //   * a boolean indicating if there is a `ref self` parameter.
+    pub fn check_self_parameter(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param_list: ast::ParamList,
+    ) -> (bool, bool) {
+        let mut add_self = true;
+        let mut has_ref_self = false;
+        if !param_list.elements(db).is_empty() {
+            let (param_name, param_modifiers, param_type) =
+                self.get_parameter_info(db, param_list.elements(db)[0].clone());
+
+            if param_name.eq(&"self".to_string()) {
+                if param_modifiers.contains(&"ref".to_string())
+                    && param_type.eq(&"ContractState".to_string())
+                {
+                    has_ref_self = false;
+                    add_self = false;
+                }
+
+                if param_type.eq(&"@ContractState".to_string()) {
+                    add_self = false;
+                }
+            }
+        };
+
+        (add_self, has_ref_self)
+    }
+
+    /// Check if the function has multiple IWorldDispatcher parameters.
+    ///
+    /// Returns
+    ///  * a boolean indicating if the function has multiple world dispatchers.
+    pub fn check_world_dispatcher(
+        &mut self,
+        db: &dyn SyntaxGroup,
+        param_list: ast::ParamList,
+    ) -> bool {
+        let mut count = 0;
+
+        param_list.elements(db).iter().for_each(|param| {
```

Comment:
What about being more strict, which may improve user comprehension. If `self` is present, `world` must be at index 1. At index `0` otherwise.

I've testing this:
```rust
fn move(direction: Direction, world: IWorldDispatcher) ...
```
And the error is:
```
error: The number of parameters in the impl function `ActionsImpl::move` is incompatible with `IActions::move`. Expected: 2, actual: 3.
 --> /dojo/examples/spawn-and-move/src/actions.cairo[actions]:99:17
        fn move(self: @ContractState, direction: Direction, world: IWorldDispatcher) {
                ^*****************************************************************^

error: Wrong number of arguments. Expected 2, found: 3
 --> /dojo/examples/spawn-and-move/src/actions.cairo[actions][contract]:209:5
    ActionsImpl::move(@contract_state, __arg_direction, __arg_world);
    ^**************************************************************^

error: Compilation failed.
```
Which can be easier for the user with a diagnostic saying: `If the world dispatcher is used, it is expected to the first parameter.`.

We've to see about `self` if we need to mention it, because user should not have to use it anyway.

Reply:
Yes, I agree !

Because it could be confusing for the user:
- if the user adds `world` as first parameter, he **must not** add it in the dojo::interface as it will be removed,
- if the user adds `world` anywhere in the function but not as first parameter, he **must** add it in the dojo::interface to avoid the compilation error you mentioned.

In fact, this is an acceptable constraint because we have the same kind of constraint for `self` when writing a smart contract :)

Reply:
But what about backward compatibility ? :-/